### PR TITLE
[ENH] Feature/datasource to xnat dag

### LINKED ISSUE #1: ENH: Creating an SFTP to XNAT DAG
I just posted the example DAG I was playing around with for getting data from the SFTP server and uploading to XNAT for SPIN-ASD. It's very basic atm. But one thing to note is that once you place a DAG in the `dag` folder, it's automatically visible in the Airflow UI.

Next steps if continuing on with this DAG are:
- [x] look into the Airflow SFTP sensor - use it to ping the SFTP server (only run on active studies)
- [x] if new files are detected, run the rest of the workflow
- [ ] credentials are stored in the Airflow database - are we ok with this?

### PR DESCRIPTION
**Pull request by jerdra**
Resolves #1

### CODE REVIEW

**Comment by josephmje** [comment]
@jerdra Sorry for the delay in reviewing this. I'm just getting around to reading about testing with Airflow so it may take me a few more days. For now, I had some questions about the project structure.

I created a mockup below of something that felt a little more intuitive to me. WTDT?

Would it make sense to move the `dags` folder outside of the `airtigrs` folder and have `airtigrs` just contain the custom code that we would want to import? To me, the `dags` folder feels like a `scripts` folder you might see in other python packages.

I also thought of combining all of the utility functions into a single folder and moving the test data to `data/tests`.

```
├── README.md
├── airtigrs
│   ├── __init__.py
│   ├── data
│   │   └── tests
│   │       ├── TEST_settings.yml
│   │       └── test_config.yaml
│   ├── operators
│   │   ├── __init__.py
│   │   └── sftp.py
│   ├── tests
│   │   ├── __init__.py
│   │   ├── test_datasources_to_xnat.py
│   │   ├── test_sftp_operator.py
│   │   └── utils.py
│   └── utils
│       ├── __init__.py
│       ├── connections.py
│       ├── scheduling.py
│       └── xnat.py
├── dags
│   └── datasources_to_xnat.py
└── docs
    └── assets
        └── images
```

**Comment by jerdra** [comment]
Thanks for the feedback @josephmje, I agree that the current directory structure is a bit messy. Could we make a minor modification to your proposal?

The test data is specifically for testing DAG building and is not expected to be used for testing operators/sensors/hooks, so maybe it should live alongside there instead of in the package. With having DAGs outside of `airtigrs` would mean that airtigrs only provides airflow-datman plugins and utilities

**Comment by josephmje** [comment]
> The test data is specifically for testing DAG building and is not expected to be used for testing operators/sensors/hooks, so maybe it should live alongside there instead of in the package. That way it would serve a role like integration tests vs. unit tests

That sounds good to me.

Also, RE the organization of tests, I've seen people do it by either having a single `tests/` folder or by having multiple `tests/` folders within each subfolder of code.

Something like:
```
└── tests/
    ├── dags/
    │   └── test_datasources_to_xnat.py
    └── airtigrs/
        └── operators/
            └── test_sftp_operator.py
        └── utils/
            └── test_connections.py
```
VS
```
├── airtigrs/
│   ├── operators/
│   │   ├── __init__.py
│   │   ├── sftp.py
│   │   └── tests/
│   │       ├── __init__.py
│   │       ├── test_sftp_operator.py
│   │       └── utils.py
│   └── utils/
│       ├── __init__.py
│       ├── connections.py
│       ├── scheduling.py
│       ├── xnat.py
│       └── tests
└── dags
    ├── datasources_to_xnat.py
    └── tests
        └── test_datasources_to_xnat.py
```

I think the 2nd one (including tests in their respective subfolder) feels a bit more natural to me - but no huge preference.

**Comment by jerdra** [comment]
haha i was also going to propose that but decided maybe to bring it up later when we build on what we have. I like having tests within the submodule because it makes tests more visible during development rather than as a side-note that is often forgotten

**Comment by jerdra** [comment]
Tagging @DESm1th @slimnsour @benselby for opinions

**Comment by jerdra** [comment]
Current directory structure is the following:

```
├── airtigrs
│   ├── __init__.py
│   ├── operators
│   │   ├── __init__.py
│   │   ├── sftp.py
│   │   └── tests
│   │       ├── __init__.py
│   │       └── test_sftp.py
│   └── utils
│       ├── connections.py
│       ├── dates.py
│       ├── tests
│       └── xnat.py
├── dags
│   ├── datasources_to_xnat.py
│   └── tests
│       ├── __init__.py
│       ├── test_datasources_to_xnat
│       │   ├── test_config.yaml
│       │   └── TEST_settings.yml
│       ├── test_datasources_to_xnat.py
│       └── utils.py
└── README.md
```

We can move `dags/tests/datasources_to_xnat/*` to a `common/` fixture directory if we end up needing the same config across multiple tests?

**Comment by josephmje** [comment]
> We can move `dags/tests/datasources_to_xnat/*` to a `common/` fixture directory if we end up needing the same config across multiple tests?

Yup, I anticipate the yml files will be used across DAGs

**Comment by jerdra** [comment]
okay so i refactored the structure a bit but it ended up complicating the testing a bit since the dag uses a factory method that we can't really test since its no longer importable.

Some suggestions to handle this:
1. Move DAG back into airtigrs making datasources_to_xnat.py importable
2. Pull out the make_dag function into a dag_factory module within air-tigrs, this would make testing `make_dag` a lot easier as well.
3. Remove testing DAG configuration entirely and instead use integration tests in CI. 

I kinda prefer #3 since I feel like with complex DAGs you don't really wanna test every possible configuration. Having some benchmark expectations feels alright.

#2 is my second choice since it'll make `make_dag` easily testable without unintentionally running unwrapped code in `datasources_to_xnat.py`

**Comment by josephmje** [comment]
> 3. Remove testing DAG configuration entirely and instead use integration tests in CI.

I think I'm also with you on point 3. Atm, our pipelines don't quite conform to the DAG dependency nature since they will still run even if out of order. I think the more important things to test might be DAG integrity or whether we have cyclical tasks: https://medium.com/wbaa/datas-inferno-7-circles-of-data-testing-hell-with-airflow-cef4adff58d8

File: airtigrs/tests/test_sftp_operator.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+import pysftp
+
+from airtigrs.plugins.operators.sftp import SFTPFetchOperator
+from airtigrs.utils import as_datetime
+from airflow.models import Connection
+
+from airflow.exceptions import AirflowSkipException
+
+DATA_INTERVAL_START = datetime.datetime(2021, 9, 13, tzinfo=pytz.UTC)
+DATA_INTERVAL_END = DATA_INTERVAL_START + datetime.timedelta(days=1)
```
[the remark this answers is not in the document]
Comment (jerdra) [reply airtigrs/tests/test_sftp_operator.py:14]: good catch they can be removed!

**Comment by jerdra** [comment]
> > 3. Remove testing DAG configuration entirely and instead use integration tests in CI.
> 
> I think I'm also with you on point 3. Atm, our pipelines don't quite conform to the DAG dependency nature since they will still run even if out of order. I think the more important things to test might be DAG integrity or whether we have cyclical tasks: https://medium.com/wbaa/datas-inferno-7-circles-of-data-testing-hell-with-airflow-cef4adff58d8

That's a nice article! Agreed that testing each module for cycles is a good strategy. I found this resource to be very helpful:

https://airflowsummit.org/slides/j2-Ensuring-your-DAGs-work-before-going-to-production.pdf

**Comment by jerdra** [comment]
@josephmje @slimnsour @benselby @DESm1th let's get this PR merged in! 

Since we've decided w/Option 3 (i.e removing DAG unit tests in favour of testing in CI) I'll go ahead removing the full DAG tests as it is currently implemented.

Instead in a future set of PR's (will make issues) we'll do the following:

1. Implement the current unit testing using github actions
2. Refactor out our current DAG into a factory module (as discussed during our air-tigrs meeting), it'll be more testable this way w/functions being importable
3. Then implement full DAG integrity tests in CI

**Review by josephmje — APPROVED** [review]

**Review by slimnsour — APPROVED** [review]
Sorry for not posting my review sooner, I'm with you guys on choosing option 3 though. Thanks so much for the hard work!

**Pull request merged**